Homework for sprint №16 (MVVM)

sobolevkir opened the pull request:
1. Произведён полный рефакторинг приложения в соотстветствии с чистойархитектурой.
2. Использован шаблон MVVM.
3. Добавлена фича - любимые треки. При поиске производится сортировка треков в списке результатов. Первыми показываются любимые треки.

P.S.: У меня просьба, можете пожалуйста посмотреть, есть сомнение, что интеракторы можно передавать в репозитории - это наверное не совсем правильно. Но я не нашёл другого способа использовать фичу Favorites в поиске треков, чтобы отсортировать результат...

sobolevkir committed:
1. Произведён полный рефакторинг приложения в соотстветствии с чистой архитектурой.
2. Использован шаблон MVVM.
3. Добавлена фича - любимые треки. При поиске производится сортировка треков в списке результатов. Первыми показываются любимые треки.

nadzyathedude commented:
`У меня просьба, можете пожалуйста посмотреть, есть сомнение, что интеракторы можно передавать в репозитории - это наверное не совсем правильно. Но я не нашёл другого способа использовать фичу Favorites в поиске треков, чтобы отсортировать результат...` - простой ответ - эту фичу мы сделаем в следующих спринтах другим подходом, пока что можешь убрать это из своего кода

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/player/data/impl/PlayerImpl.kt:
```diff
@@ -0,0 +1,67 @@
+    private var playerState = PlayerState.DEFAULT
+
+    override fun preparePlayer(consumer: Player.Consumer) {
+        Log.d("player", "preparePlayer()")
```
логирование стоит всегда удалять. во-первых это правильо чистого кода, во-вторых это может привести к утечке логов. Можно использовать Timber. Он почти то же самое, что Log, немного больше в нем функционала, но он дает безопасность
вот хорошая [статья](https://medium.com/mindorks/better-logging-in-android-using-timber-72e40cc2293d), только сейчас сайт medium открывается с впн

sobolevkir replied:
принято! 🫡

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/player/ui/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,123 @@
+        viewModel.getProgressLiveData().observe(this) { progress ->
+            binding.tvCurrentPosition.text = progress
+        }
+        with(binding) {
+            btnBack.setOnClickListener { finish() }
+            btnPlayControl.setOnClickListener { viewModel.playbackControl() }
+            btnFavorite.setOnClickListener { viewModel.onFavoriteButtonClick() }
+        }
```
это можно вынести смело в отдельную функцию, назвать например `initListeners()`

sobolevkir replied:
принято! 🫡

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/player/ui/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,123 @@
+    private fun setTrackInfo(track: Track) {
+        with(binding) {
+            btnFavorite.setImageDrawable(getFavoriteButtonDrawable(track.isFavorite))
+            Glide.with(root.context).load(track.artworkUrl512).centerInside()
```
тут смело можно использовать контекст активити

sobolevkir replied:
> тут смело можно использовать контекст активити

что именно рекомендуется сделать? вместо root.context что-то другое написать?)

nadzyathedude replied:
context@this - где this это активити

sobolevkir replied:
а в чём отличие от root.context?

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/search/data/impl/TracksRepositoryImpl.kt:
```diff
@@ -0,0 +1,67 @@
+        val response = networkClient.doRequest(TracksSearchRequest(searchQueryText))
+        return when (response.resultCode) {
+            -1 -> Resource.Error(ErrorType.CONNECTION_PROBLEM)
+            200 -> {
+                if ((response as TracksSearchResponse).results.isEmpty()) {
+                    Resource.Error(ErrorType.NOTHING_FOUND)
+                } else {
+                    Resource.Success(TrackMapper.map(response.results, favoritesInteractor.getSavedFavoritesId()))
+                }
+            }
+
+            400 -> Resource.Error(ErrorType.BAD_REQUEST)
+            404 -> Resource.Error(ErrorType.NOTHING_FOUND)
```
предлагаю вынести резолт коды ошибок к константы и например прям в отдельным класс

sobolevkir replied:
принято! 🫡

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/search/ui/activity/SearchActivity.kt:
```diff
@@ -0,0 +1,178 @@
+    private var searchTextWatcher: TextWatcher? = null
+    private lateinit var binding: ActivitySearchBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
очень громоздкая функция `onCreate()` вышла. она должна быть максимально читаемой в несколько строчек. вынеси все действия по смыслу в отдельные функции и вызови их в` onCreate()`

sobolevkir replied:
принято! 🫡

nadzyathedude commented on app/src/main/java/com/sobolevkir/playlistmaker/search/ui/viewmodel/SearchViewModel.kt:
```diff
@@ -0,0 +1,138 @@
+import com.sobolevkir.playlistmaker.search.ui.activity.SearchActivity
+import com.sobolevkir.playlistmaker.search.ui.model.SearchState
+
+class SearchViewModel(application: Application) : AndroidViewModel(application) {
```
чтобы ViewModel не зависила от Application, наследуй класс не от AndroidViewModel, а от класса ViewModel

sobolevkir replied:
> чтобы ViewModel не зависила от Application, наследуй класс не от AndroidViewModel, а от класса ViewModel

вот так? 
<img width="668" alt="image" src="https://github.com/sobolevkir/PlaylistMaker/assets/108625077/78be4aaf-9af5-446d-9edf-96f0f063d9cc">

sobolevkir replied:
чем это лучше?

sobolevkir replied:
нам же нужен контест в tracksRepository, как тогда  без него обойтись? У меня ещё в других ВМ есть контексты...  🤔

sobolevkir replied:
например вот здесь, чтобы сменить тему: при переключении:

<img width="598" alt="image" src="https://github.com/sobolevkir/PlaylistMaker/assets/108625077/1dcdd7a4-f9bd-4325-8484-ce8f9e960cfa">

sobolevkir replied:
Ну и у нас вроде в теории было показано, как можно передать во вьюмодел контекст..

sobolevkir replied:
Надя, подскажи пожалуйста, как лучше сделать и почему :
1. Передать Context во ViewModel, затем оттуда в Interactor, затем оттуда в Repository
2. Передать Context через Creator в Repository, минуя ViewModel и Interactor

nadzyathedude replied:
через Creator

nadzyathedude replied:
потому что тогда будет слишком много связей между слоями, а нам по чистой архитектуре как раз хочется как можно меньше связей между слоями

sobolevkir committed:
Скорректирован класс TrackRepository. Он теперь не использует интерактор FavoritesInteractor. Вместо этого в интеракторе TracksInteractor используются 2 репозитория - tracksRepository и favoritesRepository.

sobolevkir commented:
> `У меня просьба, можете пожалуйста посмотреть, есть сомнение, что интеракторы можно передавать в репозитории - это наверное не совсем правильно. Но я не нашёл другого способа использовать фичу Favorites в поиске треков, чтобы отсортировать результат...` - простой ответ - эту фичу мы сделаем в следующих спринтах другим подходом, пока что можешь убрать это из своего кода

Эту фичу поправил - теперь tracksInteractor использует помимо tracksRepository - favoritesRepository. Т.е. направление зависимостей корректно теперь.

sobolevkir committed:
1. Инициализация слушателей в активити вынесена в отдельные методы initListeners().
2. В активити плеера удален контекст из конструктора VM, т.к. он не используется.
3. Коды ответов от сервера вынесены в класс с константами ResultCode.
4. Удалены лишние логи.

sobolevkir committed:
1. Все ViewModel теперь не используют контекст приложения, а только интеракторы.
2. Логика по работе с темой перемещена из класса App в SettingsRepository, а в самом классе App работа с темой производится через SettingsInteractor.
3. Creator подвергся рефакторингу - контекст application передаётся теперь в репозитории там, где это необходимо.

sobolevkir committed:
Скорректированы зависимости. Интеракторы теперь передаются в конструкторы ViewModel.
